code/modules/mob/living/basic/ruin_defender/mimic/mimic.dm

```diff
@@ -0,0 +1,381 @@
+		else if(isitem(original))
+			var/obj/item/I = original
+			health = 15 * I.w_class
+			melee_damage_lower = 2 + I.force
+			melee_damage_upper = 2 + I.force
+		maxHealth = health
+		if(user)
+			befriend(user)
+		if(destroy_original)
+			qdel(original)
+		return TRUE
+	return FALSE
+
+/// Copies the object visually including name and desc
+/mob/living/basic/mimic/copy/proc/CopyObjectVisuals(obj/original)
+	name = original.name
+	desc = original.desc
+	icon = original.icon
+	icon_state = original.icon_state
+	icon_living = icon_state
+	copy_overlays(original)
+
+/mob/living/basic/mimic/copy/machine
+	ai_controller = /datum/ai_controller/basic_controller/mimic_copy/machine
+	faction = list(FACTION_MIMIC, FACTION_SILICON)
+
+/mob/living/basic/mimic/copy/ranged
+	icon = 'icons/turf/floors.dmi'
+	icon_state = "invisible"
+	ai_controller = /datum/ai_controller/basic_controller/mimic_copy/gun
+	/// the gun we are
+	var/obj/item/gun/gun
+
+/mob/living/basic/mimic/copy/ranged/Destroy()
+	vis_contents.Cut()
+	gun = null
+	return ..()
+
+/mob/living/basic/mimic/copy/ranged/RangedAttack(atom/atom_target, modifiers)
+	INVOKE_ASYNC(src, PROC_REF(fire_gun), atom_target, modifiers)
+
+/mob/living/basic/mimic/copy/ranged/proc/fire_gun(atom/target, modifiers) // i cant find any better way to do this
+	if(!gun.can_shoot())
+		if(istype(gun, /obj/item/gun/ballistic))
+			var/obj/item/gun/ballistic/ballistic = gun
+			if(!ballistic.chambered || ballistic.bolt_locked)
+				ballistic.rack() //we racked so both checked variables should be something else now
+			// do we have nothing chambered/chambered is spent AND we have no mag or our mag is empty
+			if(!ballistic.chambered?.loaded_projectile && magazine_useless(gun)) // ran out of ammo
+				ai_controller?.set_blackboard_key(BB_GUNMIMIC_GUN_EMPTY, TRUE) //BANZAIIIIIIII
+				ai_controller?.CancelActions()
+		else //if we cant fire we probably like ran out of energy or magic charges or whatever the hell idk
+			ai_controller?.set_blackboard_key(BB_GUNMIMIC_GUN_EMPTY, TRUE)
+			ai_controller?.CancelActions() // Stop our firing behavior so we can plan melee
+	else
+		ai_controller?.set_blackboard_key(BB_GUNMIMIC_GUN_EMPTY, FALSE)
+	gun.fire_gun(target, user = src, flag = FALSE, params = modifiers) //still make like a cool click click sound if trying to fire empty
+
+/mob/living/basic/mimic/copy/ranged/proc/magazine_useless(obj/item/gun/ballistic/ballistic)
+	if(isnull(ballistic.magazine) || !length(ballistic.magazine.stored_ammo))
+		return TRUE
+	// is there ATLEAST one unspent round (for the sake of revolvers or a magazine somehow having spent rounds in it)
+	for(var/obj/item/ammo_casing/thing as anything in ballistic.magazine.stored_ammo)
+		if(ispath(thing))
+			return FALSE // unspent
+		if(!isnull(thing.loaded_projectile))
+			return FALSE //unspent
+	return TRUE
+
+/mob/living/basic/mimic/copy/ranged/CopyObject(obj/item/gun/original, mob/living/creator, destroy_original = 0)
+	gun = original
```

Comment:
We should either do a safety check when reading the value of gun in other procs, or register here signals for gun being deleted or moved and kill the mob on operation.  
I am thinking of edge cases like a wizard animating a gun they have cast the recall spell on and then summoning it back into their hand and animating it again (actually, maybe do that in every mimic 😅)